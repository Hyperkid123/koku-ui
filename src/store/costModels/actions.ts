import {
  CostModel,
  CostModelRequest,
  CostModels,
  deleteCostModel as apiDeleteCostModel,
  fetchCostModels as apiGetCostModels,
  updateCostModel as apiUpdateCostModel,
} from 'api/costModels';
import { AxiosError, AxiosResponse } from 'axios';
import * as H from 'history';
import { Dispatch } from 'react-redux';
import { createAsyncAction, createStandardAction } from 'typesafe-actions';

interface FilterQuery {
  currentFilterType?: string;
  currentFilterValue?: string;
}

export const updateFilterToolbar = createStandardAction(
  'fetch/costModels/filter'
)<FilterQuery>();

export const selectCostModel = createStandardAction('select/costModels')<
  CostModel
>();

export const resetCostModel = createStandardAction('reset/costModels')<void>();

interface DialogPayload {
  isOpen: boolean;
  name: string;
}

export const setCostModelDialog = createStandardAction(
  'display/costModels/dialog'
)<DialogPayload>();

export const {
  request: fetchCostModelsRequest,
  success: fetchCostModelsSuccess,
  failure: fetchCostModelsFailure,
} = createAsyncAction(
  'fetch/costModels/request',
  'fetch/costModels/success',
  'fetch/costModels/failure'
)<void, AxiosResponse<CostModels>, AxiosError>();

export const fetchCostModels = (query: string = '') => {
  return (dispatch: Dispatch) => {
    dispatch(fetchCostModelsRequest());

    return apiGetCostModels(query)
      .then(res => {
        dispatch(fetchCostModelsSuccess(res));
      })
      .catch(err => {
        dispatch(fetchCostModelsFailure(err));
      });
  };
};

export const {
  request: updateCostModelsRequest,
  success: updateCostModelsSuccess,
  failure: updateCostModelsFailure,
} = createAsyncAction(
  'update/costModels/request',
  'update/costModels/success',
  'update/costModels/failure'
)<void, AxiosResponse<CostModel>, AxiosError>();

export const updateCostModel = (
  uuid: string,
  request: CostModelRequest,
  dialog: string = null
) => {
  return (dispatch: Dispatch) => {
    dispatch(updateCostModelsRequest());

    return apiUpdateCostModel(uuid, request)
      .then(res => {
        dispatch(updateCostModelsSuccess(res));
        fetchCostModels()(dispatch);
        if (dialog !== null) {
          dispatch(setCostModelDialog({ name: dialog, isOpen: false }));
        }
      })
      .catch(err => {
        dispatch(updateCostModelsFailure(err));
      });
  };
};

export const {
  request: deleteCostModelsRequest,
  success: deleteCostModelsSuccess,
  failure: deleteCostModelsFailure,
} = createAsyncAction(
  'delete/costModels/request',
  'delete/costModels/success',
  'delete/costModels/failure'
)<void, void, AxiosError>();

export const deleteCostModel = (
  uuid: string,
  dialog: string = '',
  history: H.History = null
) => {
  return (dispatch: Dispatch) => {
    dispatch(deleteCostModelsRequest());

    return apiDeleteCostModel(uuid)
      .then(res => {
        dispatch(deleteCostModelsSuccess());
        dispatch(resetCostModel());
        fetchCostModels()(dispatch);
        if (dialog !== null) {
          if (dialog === 'deleteCostModel' && history) {
            history.push('/cost-models');
          }
          dispatch(setCostModelDialog({ name: dialog, isOpen: false }));
        }
      })
      .catch(err => {
        dispatch(deleteCostModelsFailure(err));
      });
  };
};
