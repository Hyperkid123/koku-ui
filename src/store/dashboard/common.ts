export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget {
  availableTabs?: any[];
  currentTab: any;
  details: {
    costKey?: string /** i18n label key */;
    formatOptions: ValueFormatOptions;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n label key */;
  };
  filter?: {
    limit?: number;
    service?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  reportType: any;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
  };
  trend: {
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems: {
    formatOptions: {};
  };
}
