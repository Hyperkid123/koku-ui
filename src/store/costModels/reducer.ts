import { CostModels } from 'api/costModels';
import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { ActionType, getType } from 'typesafe-actions';
import {
  fetchCostModelsFailure,
  fetchCostModelsRequest,
  fetchCostModelsSuccess,
  updateFilterToolbar,
} from './actions';

export const stateKey = 'costModels';

export type CostModelsState = Readonly<{
  costModels: CostModels;
  error: AxiosError;
  status: FetchStatus;
  currentFilterType: string;
  currentFilterValue: string;
}>;

export const defaultState: CostModelsState = {
  costModels: null,
  error: null,
  status: FetchStatus.none,
  currentFilterType: 'name',
  currentFilterValue: '',
};

export type CostModelsAction = ActionType<
  | typeof fetchCostModelsFailure
  | typeof fetchCostModelsRequest
  | typeof fetchCostModelsSuccess
  | typeof updateFilterToolbar
>;

export const reducer = (
  state: CostModelsState = defaultState,
  action: CostModelsAction
): CostModelsState => {
  switch (action.type) {
    case getType(fetchCostModelsRequest):
      return {
        ...state,
        status: FetchStatus.inProgress,
      };
    case getType(fetchCostModelsSuccess):
      return {
        ...state,
        currentFilterValue: '',
        status: FetchStatus.complete,
        error: null,
        costModels: action.payload.data,
      };
    case getType(fetchCostModelsFailure):
      return {
        ...state,
        status: FetchStatus.complete,
        error: action.payload,
      };
    case getType(updateFilterToolbar):
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};
