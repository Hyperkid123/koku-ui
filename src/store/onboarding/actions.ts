import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { addProvider as apiCreateProvider } from 'api/providers';
import { Provider, ProviderRequest } from 'api/providers';
import { AxiosError } from 'axios';
import i18next from 'i18next';
import { Dispatch } from 'react-redux';
import { sourcesActions } from 'store/sourceSettings';
import {
  createAction,
  createAsyncAction,
  createStandardAction,
} from 'typesafe-actions';
import { Validator } from 'utils/validators';

export const updateType = createStandardAction('onboarding/update/type')<
  string,
  Validator
>();

export const updateName = createStandardAction('onboarding/update/name')<
  string,
  Validator
>();

export const updateClusterID = createStandardAction(
  'onboarding/update/clusterId'
)<string, Validator>();

export const updateS3BucketName = createStandardAction(
  'onboarding/update/s3BucketName'
)<string, Validator>();

export const updateArn = createStandardAction('onboarding/update/ARN')<
  string,
  Validator
>();

interface CheckPayload {
  item: string;
  value: boolean;
}

export const updateSourceKindCheckList = createStandardAction(
  'onboarding/update/sourceKindCheck'
)<CheckPayload>();

export const checkSourceKindCheckList = createAction(
  'onboarding/checkAll/sourceKindCheck'
);

export const openModal = createAction('onboarding/modal/open');

export const closeModal = createAction('onboarding/modal/close');

export const cancelOnboarding = createAction('onboarding/cancel');

export const {
  request: addSourceRequest,
  success: addSourceSuccess,
  failure: addSourceFailure,
} = createAsyncAction(
  'onboarding/source/add/request',
  'onboarding/source/add/success',
  'onboarding/source/add/failure'
)<void, Provider, AxiosError>();

export function addSource(request: ProviderRequest, sideEffect?: () => void) {
  return (dispatch: Dispatch) => {
    dispatch(addSourceRequest());
    return apiCreateProvider(request)
      .then(response => {
        dispatch(addSourceSuccess(response.data));
        sourcesActions.fetchSources()(dispatch);
        dispatch(
          addNotification({
            title: i18next.t('onboarding.notification.title', {
              name: request.name,
            }),
            description: i18next.t('onboarding.notification.description'),
            dismissable: true,
            variant: 'success',
          })
        );
        if (sideEffect !== null) {
          sideEffect();
        }
      })
      .catch(err => {
        dispatch(addSourceFailure(err));
      });
  };
}
