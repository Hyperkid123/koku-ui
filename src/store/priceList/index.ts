import * as priceListActions from './actions';
import {
  Action as PirceListAction,
  reducer as priceListReducer,
  State as PriceListState,
  stateKey as priceListStateKey,
} from './reducer';
import * as priceListSelectors from './selectors';

export {
  priceListActions,
  priceListStateKey,
  PirceListAction,
  PriceListState,
  priceListReducer,
  priceListSelectors,
};
