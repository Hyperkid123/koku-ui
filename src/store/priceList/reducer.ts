import { Rates } from 'api/rates';
import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { ActionType, getType } from 'typesafe-actions';
import {
  fetchPriceListFailure,
  fetchPriceListRequest,
  fetchPriceListSuccess,
} from './actions';

export const stateKey = 'priceList';

export type State = Readonly<{
  rates: Rates;
  error: AxiosError;
  status: FetchStatus;
}>;

export const defaultState: State = {
  rates: null,
  error: null,
  status: FetchStatus.none,
};

export type Action = ActionType<
  | typeof fetchPriceListRequest
  | typeof fetchPriceListSuccess
  | typeof fetchPriceListFailure
>;

export function reducer(state = defaultState, action: Action): State {
  switch (action.type) {
    case getType(fetchPriceListRequest):
      return {
        ...state,
        status: FetchStatus.inProgress,
      };
    case getType(fetchPriceListSuccess):
      return {
        ...state,
        rates: action.payload,
        error: null,
        status: FetchStatus.complete,
      };
    case getType(fetchPriceListFailure):
      return {
        ...state,
        error: action.payload,
        status: FetchStatus.complete,
      };
    default:
      return state;
  }
}
