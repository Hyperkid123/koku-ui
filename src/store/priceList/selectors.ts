import { RootState } from 'store/rootReducer';
import { stateKey } from './reducer';

export const getRateTierTimeRange = (unit: string) => {
  const sep = unit.split('-');
  if (sep.length === 2) {
    return { unit: sep[0], range: sep[1] };
  }
  return { unit: null, period: null };
};

export const rateFlatter = rate => {
  return rate.tiered_rate.map((tier, ix) => {
    const time_range = getRateTierTimeRange(tier.usage.unit);
    return {
      provider_uuids: rate.provider_uuids,
      display: rate.metric.display_name,
      metric_type: rate.metric.name,
      index: ix,
      value: tier.value,
      value_unit: tier.unit,
      range_value: [tier.usage.usage_start, tier.usage.usage_end],
      range_unit: time_range.unit,
      period: time_range.range,
    };
  });
};

export const priceList = (state: RootState) => state[stateKey];

export const rates = (state: RootState) =>
  priceList(state).rates && priceList(state).rates.data;

export const ratesPerProvider = (state: RootState) =>
  rates(state) &&
  rates(state)
    .map(rateFlatter)
    .reduce((acc, rateArray) => {
      return [...acc, ...rateArray];
    }, [])
    .reduce((acc, rate) => {
      let next = { ...acc };
      rate.provider_uuids.forEach(uuid => {
        const prev = acc[uuid] || [];
        next = {
          ...next,
          [uuid]: [...prev, rate],
        };
      });
      return next;
    }, {});

export const status = (state: RootState) => priceList(state).status;

export const error = (state: RootState) => priceList(state).error;
