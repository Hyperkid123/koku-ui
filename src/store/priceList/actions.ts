import { fetchRate } from 'api/rates';
import { Rates } from 'api/rates';
import { AxiosError } from 'axios';
import { Dispatch } from 'react-redux';
import { expirationMS, FetchStatus } from 'store/common';
import { RootState } from 'store/rootReducer';
import { createStandardAction } from 'typesafe-actions';

import { cachedRates, status } from './selectors';

interface Meta {
  providerUuid: string;
}

export const fetchPriceListRequest = createStandardAction('priceList/request')<Meta>();

export const fetchPriceListSuccess = createStandardAction('priceList/success')<Rates, Meta>();

export const fetchPriceListFailure = createStandardAction('priceList/failure')<AxiosError, Meta>();

function isExpired(state: RootState, meta: Meta) {
  const cachedData = cachedRates(state, meta.providerUuid);
  const reqStatus = status(state, meta.providerUuid);
  if (reqStatus === FetchStatus.inProgress) {
    return false;
  }

  if (!cachedData) {
    return true;
  }

  const now = Date.now();
  return now > cachedData.timeRequested + expirationMS;
}

export function fetchPriceList(providerUuid) {
  const meta = { providerUuid };
  return (dispatch: Dispatch, getState) => {
    if (!isExpired(getState(), meta)) {
      return;
    }
    dispatch(fetchPriceListRequest(meta));
    return fetchRate(providerUuid)
      .then(res => {
        dispatch(fetchPriceListSuccess(res.data, meta));
      })
      .catch(err => {
        dispatch(fetchPriceListFailure(err, meta));
      });
  };
}
