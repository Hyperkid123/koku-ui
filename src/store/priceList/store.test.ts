jest.mock('api/rates');

import { fetchRate, Rates } from 'api/rates';
import { FetchStatus } from 'store/common';
import { createMockStoreCreator } from 'store/mockStore';
import * as actions from './actions';
import { reducer, stateKey } from './reducer';
import * as selectors from './selectors';

const createProdvidersStore = createMockStoreCreator({
  [stateKey]: reducer,
});

const fetchMock = fetchRate as jest.Mock;

const fixture = [
  {
    uuid: 'a1',
    provider_uuid: 'p1',
    metric: 'cpu_usage_per_hour',
    tiered: {
      unit: 'usd',
      value: 2.03,
    },
  },
];

const ratesMock: Rates = {
  count: fixture.length,
  first: 'link1',
  previous: null,
  next: 'link2',
  last: 'link2',
  data: fixture,
};

test('test fetch price list action', async () => {
  fetchMock
    .mockReturnValueOnce(Promise.resolve({ data: ratesMock }))
    .mockReturnValueOnce(Promise.reject(Error('rejected!')));
  const store = createProdvidersStore();
  await store.dispatch(actions.fetchPriceList());
  expect(fetchMock).toBeCalled();
  expect(selectors.rates(store.getState())).toEqual(ratesMock.data);
  await store.dispatch(actions.fetchPriceList());
  expect(fetchMock).toBeCalled();
  expect(selectors.error(store.getState())).toEqual(Error('rejected!'));
});

test('default state', async () => {
  const store = createProdvidersStore();
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('from init to fetch request', async () => {
  const store = createProdvidersStore();
  expect(selectors.status(store.getState())).toBe(FetchStatus.none);
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('from fetch request to fetch success', async () => {
  const store = createProdvidersStore();
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
  store.dispatch(actions.fetchPriceListSuccess(ratesMock));
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('from fetch request to fetch error', async () => {
  const store = createProdvidersStore();
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
  store.dispatch(actions.fetchPriceListFailure(Error('Opps!')));
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('from fetch success to fetch request', async () => {
  const store = createProdvidersStore();
  store.dispatch(actions.fetchPriceListSuccess(ratesMock));
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('from fetch error to fetch request', async () => {
  const store = createProdvidersStore();
  store.dispatch(actions.fetchPriceListFailure(Error('Opps!')));
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.priceList(store.getState())).toMatchSnapshot();
});

test('test selectors', async () => {
  const store = createProdvidersStore();
  store.dispatch(actions.fetchPriceListFailure(Error('Opps!')));
  expect(selectors.error(store.getState())).toEqual(Error('Opps!'));
  store.dispatch(actions.fetchPriceListSuccess(ratesMock));
  expect(selectors.rates(store.getState())).toEqual(ratesMock.data);
  store.dispatch(actions.fetchPriceListRequest());
  expect(selectors.status(store.getState())).toBe(FetchStatus.inProgress);
});
