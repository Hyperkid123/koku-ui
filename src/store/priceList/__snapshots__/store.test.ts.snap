// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
Object {
  "error": null,
  "rates": null,
  "status": 0,
}
`;

exports[`from fetch error to fetch request 1`] = `
Object {
  "error": [Error: Opps!],
  "rates": null,
  "status": 2,
}
`;

exports[`from fetch error to fetch request 2`] = `
Object {
  "error": [Error: Opps!],
  "rates": null,
  "status": 1,
}
`;

exports[`from fetch request to fetch error 1`] = `
Object {
  "error": null,
  "rates": null,
  "status": 1,
}
`;

exports[`from fetch request to fetch error 2`] = `
Object {
  "error": [Error: Opps!],
  "rates": null,
  "status": 2,
}
`;

exports[`from fetch request to fetch success 1`] = `
Object {
  "error": null,
  "rates": null,
  "status": 1,
}
`;

exports[`from fetch request to fetch success 2`] = `
Object {
  "error": null,
  "rates": Object {
    "count": 2,
    "data": Array [
      Object {
        "metric": Object {
          "display_name": "Volume request rate",
          "name": "storage_gb_request_per_month",
          "unit": "GB-months",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
          "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": null,
            },
            "value": 0.25,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": 1120,
              "usage_start": 500,
            },
            "value": 0.25,
          },
        ],
      },
      Object {
        "metric": Object {
          "display_name": "Compute request rate",
          "name": "cpu_core_request_per_hour",
          "unit": "core-hours",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "core-hours",
              "usage_end": 5,
              "usage_start": null,
            },
            "value": 0.2,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": 3,
            },
            "value": 0.25,
          },
        ],
      },
    ],
    "first": "link1",
    "last": "link2",
    "next": "link2",
    "previous": null,
  },
  "status": 2,
}
`;

exports[`from fetch success to fetch request 1`] = `
Object {
  "error": null,
  "rates": Object {
    "count": 2,
    "data": Array [
      Object {
        "metric": Object {
          "display_name": "Volume request rate",
          "name": "storage_gb_request_per_month",
          "unit": "GB-months",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
          "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": null,
            },
            "value": 0.25,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": 1120,
              "usage_start": 500,
            },
            "value": 0.25,
          },
        ],
      },
      Object {
        "metric": Object {
          "display_name": "Compute request rate",
          "name": "cpu_core_request_per_hour",
          "unit": "core-hours",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "core-hours",
              "usage_end": 5,
              "usage_start": null,
            },
            "value": 0.2,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": 3,
            },
            "value": 0.25,
          },
        ],
      },
    ],
    "first": "link1",
    "last": "link2",
    "next": "link2",
    "previous": null,
  },
  "status": 2,
}
`;

exports[`from fetch success to fetch request 2`] = `
Object {
  "error": null,
  "rates": Object {
    "count": 2,
    "data": Array [
      Object {
        "metric": Object {
          "display_name": "Volume request rate",
          "name": "storage_gb_request_per_month",
          "unit": "GB-months",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
          "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": null,
            },
            "value": 0.25,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": 1120,
              "usage_start": 500,
            },
            "value": 0.25,
          },
        ],
      },
      Object {
        "metric": Object {
          "display_name": "Compute request rate",
          "name": "cpu_core_request_per_hour",
          "unit": "core-hours",
        },
        "provider_uuids": Array [
          "92e9b353-100e-4e38-91bd-e0143f766130",
        ],
        "tiered_rate": Array [
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "core-hours",
              "usage_end": 5,
              "usage_start": null,
            },
            "value": 0.2,
          },
          Object {
            "unit": "USD",
            "usage": Object {
              "unit": "GB-months",
              "usage_end": null,
              "usage_start": 3,
            },
            "value": 0.25,
          },
        ],
      },
    ],
    "first": "link1",
    "last": "link2",
    "next": "link2",
    "previous": null,
  },
  "status": 1,
}
`;

exports[`from init to fetch request 1`] = `
Object {
  "error": null,
  "rates": null,
  "status": 0,
}
`;

exports[`from init to fetch request 2`] = `
Object {
  "error": null,
  "rates": null,
  "status": 1,
}
`;

exports[`rateFlatter flats tiers into array 1`] = `
Array [
  Object {
    "display": "Volume request rate",
    "index": 0,
    "metric_type": "storage_gb_request_per_month",
    "period": "months",
    "provider_uuids": Array [
      "92e9b353-100e-4e38-91bd-e0143f766130",
      "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
    ],
    "range_unit": "GB",
    "range_value": Array [
      null,
      null,
    ],
    "value": 0.25,
    "value_unit": "USD",
  },
  Object {
    "display": "Volume request rate",
    "index": 1,
    "metric_type": "storage_gb_request_per_month",
    "period": "months",
    "provider_uuids": Array [
      "92e9b353-100e-4e38-91bd-e0143f766130",
      "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
    ],
    "range_unit": "GB",
    "range_value": Array [
      500,
      1120,
    ],
    "value": 0.25,
    "value_unit": "USD",
  },
]
`;

exports[`rateFlatter flats tiers into array 2`] = `
Array [
  Object {
    "display": "Compute request rate",
    "index": 0,
    "metric_type": "cpu_core_request_per_hour",
    "period": "hours",
    "provider_uuids": Array [
      "92e9b353-100e-4e38-91bd-e0143f766130",
    ],
    "range_unit": "core",
    "range_value": Array [
      null,
      5,
    ],
    "value": 0.2,
    "value_unit": "USD",
  },
  Object {
    "display": "Compute request rate",
    "index": 1,
    "metric_type": "cpu_core_request_per_hour",
    "period": "months",
    "provider_uuids": Array [
      "92e9b353-100e-4e38-91bd-e0143f766130",
    ],
    "range_unit": "GB",
    "range_value": Array [
      3,
      null,
    ],
    "value": 0.25,
    "value_unit": "USD",
  },
]
`;

exports[`ratesPerProvider selector returns rates array by provider_uuid 1`] = `
Object {
  "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d": Array [
    Object {
      "display": "Volume request rate",
      "index": 0,
      "metric_type": "storage_gb_request_per_month",
      "period": "months",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
        "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
      ],
      "range_unit": "GB",
      "range_value": Array [
        null,
        null,
      ],
      "value": 0.25,
      "value_unit": "USD",
    },
    Object {
      "display": "Volume request rate",
      "index": 1,
      "metric_type": "storage_gb_request_per_month",
      "period": "months",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
        "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
      ],
      "range_unit": "GB",
      "range_value": Array [
        500,
        1120,
      ],
      "value": 0.25,
      "value_unit": "USD",
    },
  ],
  "92e9b353-100e-4e38-91bd-e0143f766130": Array [
    Object {
      "display": "Volume request rate",
      "index": 0,
      "metric_type": "storage_gb_request_per_month",
      "period": "months",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
        "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
      ],
      "range_unit": "GB",
      "range_value": Array [
        null,
        null,
      ],
      "value": 0.25,
      "value_unit": "USD",
    },
    Object {
      "display": "Volume request rate",
      "index": 1,
      "metric_type": "storage_gb_request_per_month",
      "period": "months",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
        "7fb3bbfb-79d2-4bf7-bd80-ec2bd665887d",
      ],
      "range_unit": "GB",
      "range_value": Array [
        500,
        1120,
      ],
      "value": 0.25,
      "value_unit": "USD",
    },
    Object {
      "display": "Compute request rate",
      "index": 0,
      "metric_type": "cpu_core_request_per_hour",
      "period": "hours",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
      ],
      "range_unit": "core",
      "range_value": Array [
        null,
        5,
      ],
      "value": 0.2,
      "value_unit": "USD",
    },
    Object {
      "display": "Compute request rate",
      "index": 1,
      "metric_type": "cpu_core_request_per_hour",
      "period": "months",
      "provider_uuids": Array [
        "92e9b353-100e-4e38-91bd-e0143f766130",
      ],
      "range_unit": "GB",
      "range_value": Array [
        3,
        null,
      ],
      "value": 0.25,
      "value_unit": "USD",
    },
  ],
}
`;
